cmake/JuliaUtils.cmake

```diff
@@ -0,0 +1,88 @@
+#
+
+include(CMakeVarMacros)
+
+get_filename_component(jl_cmake_utils_dir
+  "${CMAKE_CURRENT_LIST_FILE}" PATH)
+
+set(jl_cmake_utils_dir "${jl_cmake_utils_dir}" CACHE "" INTERNAL FORCE)
+
+function(jl_std_fpath var path)
+  if(NOT IS_ABSOLUTE "${path}")
+    set(${var} "${path}" PARENT_SCOPE)
+    return()
+  endif()
+  file(RELATIVE_PATH src_path "${CMAKE_CURRENT_SOURCE_DIR}" "${path}")
+  file(RELATIVE_PATH bin_path "${CMAKE_CURRENT_BINARY_DIR}" "${path}")
+  string(LENGTH "${src_path}" src_len)
+  string(LENGTH "${bin_path}" bin_len)
+  if(src_len GREATER bin_len)
+    set(${var} "${bin_path}" PARENT_SCOPE)
+  else()
+    set(${var} "${src_path}" PARENT_SCOPE)
+  endif()
+endfunction()
+
+function(jl_fpath_to_target var file)
+  jl_std_fpath(std_fname "${file}")
+  string(REGEX REPLACE "[^-_.a-zA-Z0-9]" "_" std_fname "${std_fname}")
+  string(REGEX REPLACE "_+" "_" std_fname "${std_fname}")
+  cmake_utils_get_unique_name("${std_fname}" target)
+  set("${var}" "${target}" PARENT_SCOPE)
+endfunction()
+
+function(jl_rewrite_dep_list var deps)
+  set(new_deps)
+  foreach(dep ${deps})
+    if(TARGET "${dep}")
+      set(new_deps ${new_deps} "$<TARGET_FILE:${dep}>")
+    elseif(IS_ABSOLUTE "${dep}")
+      set(new_deps ${new_deps} ${dep})
+    else()
+      get_filename_component(dep "${dep}" ABSOLUTE)
+      set(new_deps ${new_deps} ${dep})
+    endif()
+  endforeach()
+  set(${var} "${new_deps}" PARENT_SCOPE)
+endfunction()
+
+function(jl_rewrite_output_list var outputs)
+  set(new_outputs)
+  foreach(output ${outputs})
+    if(IS_ABSOLUTE "${output}")
+      set(new_outputs ${new_outputs} ${output})
+    else()
+      set(new_outputs ${new_outputs} "${CMAKE_CURRENT_BINARY_DIR}/${output}")
+    endif()
+  endforeach()
+  set(${var} "${new_outputs}" PARENT_SCOPE)
+endfunction()
+
+function(jl_custom_target target outputs output_deps target_deps autodep_file)
+  jl_rewrite_dep_list(output_deps "${output_deps}")
+  jl_rewrite_output_list(outputs "${outputs}")
+  if((NOT "${autodep_file}" STREQUAL "") AND
+      (NOT IS_ABSOLUTE "${autodep_file}"))
+    set(autodep_file "${CMAKE_CURRENT_BINARY_DIR}/${autodep_file}")
+  endif()
+  add_custom_target("${target}" ALL
+    COMMAND "${CMAKE_COMMAND}"
+    "-DOUTPUTS=${outputs}" "-DOUTPUT_DEPS=${output_deps}"
+    "-DAUTODEP_FILE=${autodep_file}" "-DCOMMAND_SPEC=${ARGN}"
+    "-DCUR_BIN_DIR=${CMAKE_CURRENT_BINARY_DIR}"
+    "-DCUR_SRC_DIR=${CMAKE_CURRENT_SOURCE_DIR}"
+    -P "${jl_cmake_utils_dir}/jl_custom_target.cmake"
+    DEPENDS ${target_deps} VERBATIM)
+endfunction()
```

Comment:
> We don't need cmake to regenerate anything if we add or remove Julia files

Hmm? We do because the sysimg need to be recompiled?

> but don't try to make this PR too sophisticated to start with

I would agree with you if implementing this is more complicated in length or concept than what it currently is.

To summerize what this really is, this is a **custom cmake script** that is a **generic wrapper** running at **compile time** that does **custom timestamp checking**. and the highlighted part is what you don't like about it to the best of my knowledge

I would argue that this is the same what we are doing now and to break down my argument:

1. Custom cmake script

    I would actually be happier to replace it with shell script

2. Generic wrapper

    I'm totally fine to make it sysimg compilation only. It will make it easier to understand. I just figured that I might as well make it generic if I can but it doesn't have to.

3. Running at compile time

    We do this for the makefile system to generate many things and will still need many (either shell or cmake scripts) to replace those (or keep the old one).

4. Custom timestamp checking

    This is probably the most "sophisticated" part. I would argue that it is not so different from the logic we have for [`version_git.jl`](https://github.com/JuliaLang/julia/blob/cb775031144c5d51856b9357bdf82c8900683291/base/Makefile#L77). OK it's not checking time stamps manually but it's not so different from that.